src/context.rs

```diff
@@ -40,43 +51,53 @@ use super::types::FluentValue;
 #[allow(dead_code)]
 pub struct MessageContext<'ctx> {
     pub locales: &'ctx [&'ctx str],
-    messages: HashMap<String, ast::Message>,
-    terms: HashMap<String, ast::Term>,
-    functions: HashMap<
-        String,
-        Box<
-            'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
-        >,
-    >,
+    map: HashMap<String, Variant<'ctx>>,
 }
 
 impl<'ctx> MessageContext<'ctx> {
     pub fn new(locales: &'ctx [&'ctx str]) -> MessageContext {
         MessageContext {
             locales,
-            messages: HashMap::new(),
-            terms: HashMap::new(),
-            functions: HashMap::new(),
+            map: HashMap::new(),
         }
     }
 
     pub fn has_message(&self, id: &str) -> bool {
-        self.messages.contains_key(id)
+        self.map.get(id).map_or(false, |id| {
+            if let Variant::Message(_) = id {
+                true
+            } else {
+                false
+            }
+        })
     }
 
     pub fn get_message(&self, id: &str) -> Option<&ast::Message> {
-        self.messages.get(id)
+        self.map.get(id).and_then(|id| {
+            if let Variant::Message(ref msg) = id {
+                Some(msg)
+            } else {
+                None
+            }
+        })
     }
 
     pub fn get_term(&self, id: &str) -> Option<&ast::Term> {
-        self.terms.get(id)
+        self.map.get(id).and_then(|id| {
+            if let Variant::Term(ref term) = id {
+                Some(term)
+            } else {
+                None
+            }
+        })
     }
 
     pub fn add_function<F>(&mut self, id: &str, func: F)
     where
         F: 'ctx + Fn(&[Option<FluentValue>], &HashMap<String, FluentValue>) -> Option<FluentValue>,
     {
-        self.functions.insert(id.to_string(), Box::new(func));
+        self.map
+            .insert(id.to_string(), Variant::Function(Box::new(func)));
```

Comment:
Can you guard here against already existing entry and throw if encountered? (it probably should make `add_function` return a result)